macros/src/main/scala/offheap/internal/macros/Array.scala

```diff
@@ -276,10 +279,52 @@ trait ArrayApiCommon extends ArrayCommon {
     }
   }
 
+  def reduceLeft(op: Tree) = {
+    stabilized(c.prefix.tree) { pre =>
+      val acc = freshVar("acc", A, readElem(pre, A, q"0"))
+      val sourceIndex = freshVar("i", IntTpe, q"1")
+      val sourceLength = freshVal("len", ArraySizeTpe, readSize(pre))
+      q"""
+        if ($pre.isEmpty) ${throwUnsupportedOperation("empty.reduceLeft")}
+        else {
+          $acc
+          $sourceIndex
+          $sourceLength
+          while (${sourceIndex.symbol} < ${sourceLength.symbol}) {
+            ${acc.symbol} = ${app(op, q"${acc.symbol}", readElem(pre, A, q"${sourceIndex.symbol}"))}
+            ${sourceIndex.symbol} += 1
+          }
```

Comment:
Current implementation of `iterate` method goes through all of elements of the array, while `reduce` needs to apply given operator length-1 times. So the solution is to either refactor `iterate` to allow setting first and last value for the index or initialize the accumulator inside the loop by comparing index with zero with every iteration (which is inefficient unless JVM can optimize this case).